DOCS: fixed-wing API documentation px4-ros2-interface-lib (#24785)
PX4/PX4-Autopilot: docs/en/ros2/px4_ros2_control_interface.md

  max_heading_rate_rad_s);
```

#### Fixed-Wing Lateral and Longitudinal Setpoint (FwLateralLongitudinalSetpointType)

<Badge type="warning" text="Fixed wing only" />

::: info
This setpoint type is only supported for fixed-wing vehicles.
:::

> Use the [`px4_ros2::FwLateralLongitudinalSetpointType`](https://auterion.github.io/px4-ros2-interface-lib/classpx4__ros2_1_1FwLateralLongitudinalSetpointType.html) to directly control the lateral and longitudinal dynamics of a fixed-wing vehicle.

Reviewer: What are lateral and longitudinal dynamics? I mean a way of expressing this that a high school student would understand?

Reviewer: Further, when I see later on that I can control lateral acceleration I assume you mean I can control the rate of change of sideways movement of my vehicle. Just FMI how does a fixed wing do that - presumably this is roll, but perhaps with additional lift so you don't also sink?

Author: > Just FMI how does a fixed wing do that - presumably this is roll, but perhaps with additional lift so you don't also sink?

Yes, lateral acceleration maps to roll!

As I understand it, in PX4 the lateral (roll) and longitudinal (pitch/throttle) dynamics are decoupled. This means the roll setpoint is directly derived from the lateral acceleration setpoint, without automatic compensation for altitude loss during a turn. So, if the aircraft loses altitude while banking, it's up to the longitudinal controller to correct that—it's not handled preemptively when a lateral acceleration setpoint is given. (But maybe @sfuhrer can confirm this.)

Reviewer: Nicely done. I tweaked a bit in https://github.com/PX4/PX4-Autopilot/pull/24785/files#r2101443642

> This setpoint is streamed to the PX4 _FwLateralLongitudinalControl_ module, which decouples lateral and longitudinal inputs while ensuring that vehicle limits are respected.

Reviewer: If _FwLateralLongitudinalControl_ is a module that ends up with docs etc, we should link to it - i.e. if it appears below here https://docs.px4.io/main/en/modules/modules_main.html

Author: Ah! I do think it will appear there, yes. Added:  https://github.com/PX4/PX4-Autopilot/pull/24785/commits/aa60e891c855811ffef8a53464b6bc074b026ce0#diff-26f9676b8aff7b21773d2c818191b8bb1c189332b47256ef51c5c2e74acf507eR413

To control the vehicle, at least one lateral **and** one longitudinal setpoint must be provided:

1. Of the longitudinal inputs: either `altitude` or `height_rate` must be finite to control vertical motion.
   If both are set to `NAN`, the vehicle will maintain its current altitude.
2. Of the lateral inputs: at least one of `course`, `airspeed_direction`, or `lateral_acceleration` must be finite.

For a detailed description of the controllable parameters, please refer to message definitions ([FixedWingLateralSetpoint](../msg_docs/FixedWingLateralSetpoint.md) and [FixedWingLongitudinalSetpoint](../msg_docs/FixedWingLongitudinalSetpoint.md)).

##### Basic Usage

> In the simplest case, you can provide only a `course` and an `altitude` setpoint:

Reviewer: So just to be clear, if you do this, the vehicle will do its best to fly to the altitude you specify and the course you want, within the limits of the vehicle?

Author: Yes, exactly. Should this be clarified?

Reviewer: I've attempted to do so here. I wasn't going to, but this is a good way to state that the vehicle won't move to the setpoints aggressively, which explains why you might then use the following method.

> This setpoint type offers multiple update methods, each allowing you to specify an increasing number of setpoints. In the simplest case, you can provide a `course` and an `altitude` setpoint:

Reviewer: YOu'll note I use line breaks on sentences. This is effectively cosmetic. I prefer it though because for smaller changes the github change control makes it easier to what is happening.

```suggestion
This setpoint type offers multiple update methods, each allowing you to specify an increasing number of setpoints.
In the simplest case, you can provide a `course` and an `altitude` setpoint:
```

```cpp
const float altitude_msl = 500.F;
const float course = 0.F; // due North
_fw_lateral_longitudinal_setpoint->updateWithAltitude(altitude_msl, course);
```

> From these setpoints, PX4 will compute the required _roll angle_, _pitch angle_ and _throttle_ setpoints that are sent to lower level controllers.

Reviewer: What I want to say here is that this won't fly crazy to get to the target.

```suggestion
PX4 uses the setpoints to compute the _roll angle_, _pitch angle_ and _throttle_ setpoints that are sent to lower level controllers.
Note that the commanded flight is expected to be relatively gentle/unaggressive when using this method.
```

This is done as follows:

- Lateral control output:

  course setpoint (set by user) &rarr; airspeed direction (heading) setpoint &rarr; lateral acceleration setpoint &rarr; roll angle setpoint.
- Longitudinal control output:

  altitude setpoint (set by user) &rarr; height rate setpoint &rarr; pitch angle setpoint and throttle settings.

> Alternatively, you can provide only a course and a height rate setpoint:

Reviewer: ```suggestion
Alternatively, you can provide only a course and a `height_rate` setpoint:
```

Reviewer: Why would I want to do this?

Reviewer: I think so. Suggestion here https://github.com/PX4/PX4-Autopilot/pull/24785/files#r2101457965

```cpp
const float height_rate = 2.F;
const float course = 0.F; // due North

> _fw_lateral_longitudinal_setpoint->updateWithHeightRate(altitude_msl, course);

Reviewer: Right?
```suggestion
_fw_lateral_longitudinal_setpoint->updateWithHeightRate(height_rate, course);
```

> Alternatively, you can provide a course and a `height_rate` setpoint:

Reviewer: Attempt to capture https://github.com/PX4/PX4-Autopilot/pull/24785#discussion_r2100359900

```suggestion
Setting discrete altitude setpoints will allow PX4 to choose its own rate of ascent or descent to the target altitude.
If the speed of ascent or descent is important, or needs to be dynamically controlled, you can instead provide a `course` and `height_rate` setpoint to the `updateWithHeightRate()` method:
```

```

> To additionally control the equivalent airspeed or lateral acceleration, you can specify them as the third and fourth arguments, respectively:

Reviewer: Also when you say "equivalent airspeed or lateral acceleration" - in this example you control both, so should you say "And". I think about you referred to using NaN, so perhaps in one example mention that you can choose to just control one value by specifying Nan for the other.

  max_heading_rate_rad_s);
```

#### Fixed-Wing Lateral and Longitudinal Setpoint (FwLateralLongitudinalSetpointType)

<Badge type="warning" text="Fixed wing only" />

::: info
This setpoint type is supported for fixed-wing vehicles and for VTOLs in fixed-wing mode.
:::

Use the [`px4_ros2::FwLateralLongitudinalSetpointType`](https://auterion.github.io/px4-ros2-interface-lib/classpx4__ros2_1_1FwLateralLongitudinalSetpointType.html) to directly control the lateral and longitudinal dynamics of a fixed-wing vehicle — that is, side-to-side motion (turning/banking) and forward/vertical motion (speeding up and climbing/descending), respectively.

> This setpoint is streamed to the PX4 [_FwLateralLongitudinalControl_ module](../modules/modules_controller.md#fw_lateral_longitudinal_control), which decouples lateral and longitudinal inputs while ensuring that vehicle limits are respected.

Reviewer: I updated all the metadata so now the only error comes from this link.

Is this intended to be 

```suggestion
This setpoint is streamed to the PX4 [fw_lat_lon_control ](../modules/modules_controller.md#fw-lat-lon-control) module, which decouples lateral and longitudinal inputs while ensuring that vehicle limits are respected. 
```

http://localhost:5173/px4_user_guide/en/modules/modules_controller.html#

```cpp
const float height_rate = 2.F;
const float course = 0.F; // due North
_fw_lateral_longitudinal_setpoint->updateWithHeightRate(height_rate, course);
```

The `updateWithAltitude()` and `updateWithHeightRate()` methods allow you to additionally control the equivalent airspeed or lateral acceleration by specifying them as the third and fourth arguments, respectively:

```cpp
const float altitude_msl = 500.F;
const float course = 0.F; // due North
const float equivalent_aspd = 15.F; // m/s
const float lateral_acceleration = 2.F; // FRD, used as feedforward

_fw_lateral_longitudinal_setpoint->updateWithAltitude(altitude_msl,
  course,
  equivalent_aspd,
  lateral_acceleration);
```

The equivalent airspeed and lateral acceleration arguments are defined as `std::optional<float>`, so you can omit any of them by passing `std::nullopt`.

::: tip
If both lateral acceleration and course setpoints are provided, the lateral acceleration setpoint will be used as feedforward.
:::

##### Full Control Using the Setpoint Struct

> For full flexibility, you can create and pass a `FwLateralLongitudinalSetpoint` struct.

Reviewer: `FwLateralLongitudinalSetpoint` - link please as this isn't anywhere else. It may be that it is misnamed version of the one with the `Type` suffix, in which case, just fix the name.

![FW ROS Interaction](../../assets/middleware/ros2/px4_ros2_interface_lib/fw_lat_long_ros_interaction.svg)

##### Advanced Configuration (Optional)

You can also pass a `FwControlConfiguration` struct along with the setpoint to override default controller settings and constraints such as pitch limits, throttle limits, and target sink/climb rates.
This is intended for advanced users:

```cpp
px4_ros2::FwLateralLongitudinalSetpoint setpoint_s;

setpoint_s.withAirspeedDirection(0.F);
setpoint_s.withLateralAcceleration(2.F); // feedforward
setpoint_s.withAltitude(500.F);
setpoint_s.withEquivalentAirspeed(15.F);

px4_ros2::FwControlConfiguration config_s;

config_s.withTargetClimbRate(3.F);
config_s.withMaxAcceleration(5.F);
config_s.withThrottleLimits(0.4F, 0.6F);

_fw_lateral_longitudinal_setpoint->update(setpoint_s, config_s);
```

All configuration fields are defined as `std::optional<float>`.
Unset values will default to the PX4 configuration.
See [LateralControlConfiguration](../msg_docs/LateralControlConfiguration.md) and [FixedWingLongitudinalConfiguration](../msg_docs/LongitudinalControlConfiguration.md) for more information on configuration options.

> :::info For safety, PX4 automatically limits configuration values to stay within the vehicle’s constraints. For example, throttle overrides are clamped to remain between [`FW_THR_MIN`](../advanced_config/parameter_reference.md#FW_THR_MIN)

Reviewer: Indentation matters!!! This would render with the text as a heading.

```suggestion
:::info
For safety, PX4 automatically limits configuration values to stay within the vehicle’s constraints.
For example, throttle overrides are clamped to remain between [`FW_THR_MIN`](../advanced_config/parameter_reference.md#FW_THR_MIN)
```

  max_heading_rate_rad_s);
```

#### Fixed-Wing Lateral and Longitudinal Setpoint (FwLateralLongitudinalSetpointType)

> <Badge type="warning" text="Fixed wing only" />

Reviewer 2: And VTOLs in FW mode?

Author: I added it in the info box underneath: https://github.com/PX4/PX4-Autopilot/pull/24785/commits/280260aebd089f2c8788eb6d46f1b609bb611b64 

Or do you think it should go explicitly in the badge?

Reviewer 2: up to you, not strongly opinionated.

```cpp
const float height_rate = 2.F;
const float course = 0.F; // due North
_fw_lateral_longitudinal_setpoint->updateWithHeightRate(height_rate, course);
```

The `updateWithAltitude()` and `updateWithHeightRate()` methods allow you to additionally control the equivalent airspeed or lateral acceleration by specifying them as the third and fourth arguments, respectively:

```cpp
const float altitude_msl = 500.F;
const float course = 0.F; // due North
const float equivalent_aspd = 15.F; // m/s
const float lateral_acceleration = 2.F; // FRD, used as feedforward

_fw_lateral_longitudinal_setpoint->updateWithAltitude(altitude_msl,
  course,
  equivalent_aspd,
  lateral_acceleration);
```

The equivalent airspeed and lateral acceleration arguments are defined as `std::optional<float>`, so you can omit any of them by passing `std::nullopt`.

::: tip
If both lateral acceleration and course setpoints are provided, the lateral acceleration setpoint will be used as feedforward.
:::

##### Full Control Using the Setpoint Struct

For full flexibility, you can create and pass a `FwLateralLongitudinalSetpoint` struct.
Each field is templated with `std::optional<float>`.

::: tip

> If both course and airspeed direction are set, course is not controlled.

Reviewer 2: What is instead done? Airspeed direction control?

Author: Clarified: https://github.com/PX4/PX4-Autopilot/commit/e40de52fac2016bc993c79f5e83c180a40c6c647

Lateral acceleration is treated as feedforward if either course or airspeed direction are also finite.

> If both altitude and height rate are set, altitude is not controlled.

Reviewer 2: Also here, what happens in that case?

Author: Clarified: https://github.com/PX4/PX4-Autopilot/pull/24785/commits/e40de52fac2016bc993c79f5e83c180a40c6c647

Reviewer 2: perfect

This is intended for advanced users:

```cpp
px4_ros2::FwLateralLongitudinalSetpoint setpoint_s;

setpoint_s.withAirspeedDirection(0.F);
setpoint_s.withLateralAcceleration(2.F); // feedforward
setpoint_s.withAltitude(500.F);
setpoint_s.withEquivalentAirspeed(15.F);

px4_ros2::FwControlConfiguration config_s;

config_s.withTargetClimbRate(3.F);
config_s.withMaxAcceleration(5.F);
config_s.withThrottleLimits(0.4F, 0.6F);

_fw_lateral_longitudinal_setpoint->update(setpoint_s, config_s);
```

All configuration fields are defined as `std::optional<float>`.
Unset values will default to the PX4 configuration.
See [LateralControlConfiguration](../msg_docs/LateralControlConfiguration.md) and [FixedWingLongitudinalConfiguration](../msg_docs/LongitudinalControlConfiguration.md) for more information on configuration options.

:::warning Overrides must remain within valid system constraints.
For example, throttle limits must stay between [`FW_THR_MIN`](../advanced_config/parameter_reference.md#FW_THR_MIN).
and [`FW_THR_MAX`](../advanced_config/parameter_reference.md#FW_THR_MIN).

> PX4 will disregard any configuration overrides that do not respect system constaints.

Reviewer 2: Does it not just clamp it to the PX4 internal min/max?

Reviewer 2: I would say so!

(TODO: ADD LINK HERE)

##### Basic Usage

In the simplest case, you can provide only a course and altitude setpoint:
```cpp
const float altitude_msl = 500.F;
const float course = 0.F; // due North
_fw_lateral_longitudinal_setpoint->update(altitude_msl, course);
```
PX4 will compute:

- Lateral control output: required airspeed, airspeed direction (heading), lateral acceleration, and roll angle

- Longitudinal control output: height rate, pitch angle, and throttle settings

To additionally control the height rate, equivalent airspeed, or lateral acceleration, you can specify them as the third, fourth, and fifth arguments, respectively:

```cpp
const float altitude_msl = 500.F;
const float course = 0.F; // due North
const float equivalent_aspd = 15.F; // m/s
const float lateral_acceleration = 2.F; // FRD, used as feedforward

_fw_lateral_longitudinal_setpoint->update(altitude_msl,
  course,
  std::nullopt,
  equivalent_aspd,
  lateral_acceleration);
```

The height rate, equivalent airspeed, and lateral acceleration arguments are defined as `std::optional<float>`, so you can omit any of them by passing `std::nullopt`. In the example above, the height rate is left unset.

::: tip

> If multiple lateral inputs are set, the lower level inputs are used as feedforward. If both altitude and height rate are set, *altitude is not controlled*.

Reviewer 2: > If multiple lateral inputs are set, the lower level inputs are used as feedforward

For which combinations does that work? Maybe worth to specifically point out the possible combinations? Unless it's too many options?

:::

##### Full Control Using the Setpoint Struct

For full flexibility, you can create and pass a `FwLateralLongitudinalSetpoint` struct. Each field is also templated with `std::optional<float>`. Please consult the message definitions for valid input combinations.
(TODO: ADD LINK HERE)

```cpp
  px4_ros2::FwLateralLongitudinalSetpoint setpoint_s{};

  setpoint_s.withCourse(0.F);
  setpoint_s.withAirspeedDirection(0.2F); // feedforward
  setpoint_s.withLateralAcceleration(2.F); // feedforward
  //setpoint_s.withAltitude(500.F); // uncontrolled
  setpoint_s.withHeightRate(2.F);
  setpoint_s.withEquivalentAirspeed(15.F);

  _fw_lateral_longitudinal_setpoint->update(setpoint_s);
```

##### Advanced Configuration (Optional)

> You can also pass a `FwControlConfiguration` struct along with the setpoint to override default control constraints such as pitch limits, throttle limits, and target sink/climb rates. This is intended for advanced users:

Reviewer 2: ```suggestion
You can also pass a `FwControlConfiguration` struct along with the setpoint to override default controller settings and constraints such as pitch limits, throttle limits, and target sink/climb rates. This is intended for advanced users:
```

If you want to control an actuator that does not control the vehicle's motion, but for example a payload servo, see [below](#controlling-an-independent-actuator-servo).
:::

### Controlling a VTOL

<Badge type="warning" text="Experimental" />

> The [VTOL API](https://auterion.github.io/px4-ros2-interface-lib/classpx4__ros2_1_1Vtol.html) provides the functionality to command a transition.

Reviewer: The VTOL API links don't work yet.

Reviewer: Does this API do anything else, or is transition the only VTOL specific thing?

Author: >The VTOL API links don't work yet. 

Indeed, [that PR](https://github.com/Auterion/px4-ros2-interface-lib/pull/92) hasn't gone in yet.

> Does this API do anything else, or is transition the only VTOL specific thing?

- Command transition 
- Query the current state of the VTOL

This is intended for advanced users.

> 1. Ensure that both the `TrajectorySetpointType` and the `FwLateralLongitudinalSetpointType` are available to your mode.

Reviewer: Can you tell me where TrajectorySetpointType is defined. this doesn't show up in search other than here (in the docs)

Author: https://auterion.github.io/px4-ros2-interface-lib/classpx4__ros2_1_1TrajectorySetpointType.html 

I have also added it in the docs now

If you want to control an actuator that does not control the vehicle's motion, but for example a payload servo, see [below](#controlling-an-independent-actuator-servo).
:::

### Controlling a VTOL

<Badge type="warning" text="Experimental" />

The [VTOL API](https://auterion.github.io/px4-ros2-interface-lib/classpx4__ros2_1_1Vtol.html) provides the functionality to command a transition and query the current state of the vehicle.

> This is intended for advanced users.

Reviewer: 1. Why?
2. What is an advanced user?

This matters because I'm not an advanced user but I do want to make a transition, I need some way to work out why I should/shouldn't do this.

Perhaps word as "This API should be used with caution because ...."

Author: That's a good point. I added a small explanation: https://github.com/PX4/PX4-Autopilot/pull/24785/commits/aa60e891c855811ffef8a53464b6bc074b026ce0#diff-26f9676b8aff7b21773d2c818191b8bb1c189332b47256ef51c5c2e74acf507eR551

If you want to control an actuator that does not control the vehicle's motion, but for example a payload servo, see [below](#controlling-an-independent-actuator-servo).
:::

### Controlling a VTOL

<Badge type="warning" text="Experimental" />

The [VTOL API](https://auterion.github.io/px4-ros2-interface-lib/classpx4__ros2_1_1Vtol.html) provides the functionality to command a transition and query the current state of the vehicle.

> Use this API with caution: commanding transitions externally makes the user partially responsible for ensuring smooth and safe behavior, unlike onboard transitions (e.g. via RC switch) where PX4 handles the full process.

Reviewer: What you did here is very clear thanks. I'm just tweaking.
```suggestion
Use this API with caution!
Commanding transitions externally makes the user partially responsible for ensuring smooth and safe behavior, unlike onboard transitions (e.g. via RC switch) where PX4 handles the full process.
```

If you want to control an actuator that does not control the vehicle's motion, but for example a payload servo, see [below](#controlling-an-independent-actuator-servo).
:::

> ### Controlling a VTOL

Reviewer 2: Is this only for the VTOL specific commands, while VTOL also can be controlled with the MC and FW ones above?

Author: ah good point, how's this: https://github.com/PX4/PX4-Autopilot/pull/24785/commits/86dd7a260c6dd0a78badaac8472e155ee7ffd6ff 

I added some explanation on how to control VTOLs from an external mode, seperate from the API itself.

Reviewer 2: good for now I'd say :+1:
